File: examples/documents/includes.ts
import { Post } from "../_harness/generated/ts/Post.generated";
import { Author } from "../_harness/generated/ts/Author.generated";

// Pass `include` in a query to batch-load related records alongside the rows,
// instead of issuing a follow-up query per result. Related records ride along
// under `_related` on each row, keyed by the include's `as`.
export async function loadRelatedData() {
  // #region example
  // refersTo — each Post's Author (the `authorId` FK lives on Post):
  const posts = await Post.query({}, {
    include: [{ model: "authors", type: "refersTo", sourceField: "authorId", as: "author" }],
  });
  for (const post of posts.data as (Post & { _related?: { author?: Author } })[]) {
    console.log(post.title, post._related?.author?.name);
  }

  // hasMany — every Post that points back at each Author, newest first:
  const authors = await Author.query({}, {
    include: [{ model: "posts", type: "hasMany", foreignKey: "authorId", as: "posts", sort: { createdAt: -1 }, limit: 10 }],
  });
  for (const author of authors.data as (Author & { _related?: { posts?: Post[] } })[]) {
    console.log(author.name, (author._related?.posts ?? []).length);
  }
  // #endregion example
  return { posts: posts.data, authors: authors.data };
}

File: examples/documents/includes.swift
import JsBaoClient

// Pass `include` in a query to batch-load related records alongside the rows,
// instead of following each relationship one row at a time. Every declared
// relationship gets a typed `include<Rel>()` builder and a typed `related<Rel>`
// accessor to read the attached records.
func loadRelatedData() throws {
  // #region example
  // refersTo — each Post's Author (the `authorId` FK lives on Post):
  let posts = try Post.query(include: [Post.includeAuthor()])
  for post in posts {
    let author = post.relatedAuthor          // Author?
    print(post.title, author?.name ?? "—")
  }

  // hasMany — every Post that points back at each Author, newest first:
  let authors = try Author.query(include: [Author.includePosts(limit: 10)])
  for author in authors {
    let authored = author.relatedPosts       // [Post]
    print(author.name, authored.count)
  }
  // #endregion example
  _ = posts
  _ = authors
}
